fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -2238,27 +2238,39 @@ def _gather_optim_state(
             # param that has the optimizer state. So we handle it with the correct
             # parameter list.
             non_shared_params = cast(FullyShardedDataParallel, self._fsdp_instances[k]).non_shared_params()
+            non_shared_world_size = self._fsdp_instances[k].world_size
+            non_shared_process_group = self._fsdp_instances[k].process_group
+            # This is the world size and process group of the FSDP submodule which can be
+            # different than the parent module.
+            if non_shared_world_size != self.world_size:
+                world_size = non_shared_world_size
+                process_group = non_shared_process_group
+            else:
+                world_size = self.world_size
+                process_group = self.process_group
+
             assert (
                 len(non_shared_params) == 1
             ), f"Only flatten param or a single non-shared param is supported: len={len(non_shared_params)}"
             desired_buffer_size = non_shared_params[0]._full_param_padded.size()
             buffer = None  # for sharded tensors
             singleton_buffer = None  # for singleton tensors
             for buffer_name, t in v.items():
+
```

Comment:
it might be harder to read to an extra line right after the for statement before the if statement within it.